db/pebble/db.go

```diff
@@ -87,6 +164,192 @@ func (d *DB) WithListener(listener db.EventListener) db.DB {
+		Namespace: dbNamespace,
+		Name:      "write_delay_count",
+		Help:      "Total write delay counts due to database compaction",
+	})
+	d.writeDelayTimeMeter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: dbNamespace,
+		Name:      "write_delay_time",
+		Help:      "Total time spent in write stalls",
+	})
+	d.diskSizeGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "disk_size",
+		Help:      "Tracks the size of all of the levels of the database",
+	})
+	d.diskWriteMeter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: dbNamespace,
+		Name:      "disk_write",
+		Help:      "Measures the effective amount of data written to disk",
+	})
+	d.memCompGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "mem_comp",
+		Help:      "Tracks the amount of memory allocated for compaction",
+	})
+	d.level0CompGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "level0_comp",
+		Help:      "Tracks the number of level-zero compactions",
+	})
+	d.nonlevel0CompGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "nonlevel0_comp",
+		Help:      "Tracks the number of non level-zero compactions",
+	})
+	d.seekCompGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "seek_comp",
+		Help:      "Tracks the number of table compaction caused by read opt",
+	})
+	d.manualMemAllocGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: dbNamespace,
+		Name:      "manual_mem_alloc",
+		Help:      "Tracks the amount of non-managed memory currently allocated",
+	})
+
+	prometheus.MustRegister(
+		d.compTimeMeter,
+		d.compReadMeter,
+		d.compWriteMeter,
+		d.writeDelayCountMeter,
+		d.writeDelayTimeMeter,
+		d.diskSizeGauge,
+		d.diskWriteMeter,
+		d.memCompGauge,
+		d.level0CompGauge,
+		d.nonlevel0CompGauge,
+		d.seekCompGauge,
+		d.manualMemAllocGauge,
+	)
+
+	return d
+}
+
+func (d *DB) onCompactionBegin(info pebble.CompactionInfo) { //nolint:gocritic // Used by pebble's event listener
+	if d.activeComp == 0 {
+		d.compStartTime = time.Now()
+	}
+	l0 := info.Input[0]
+	if l0.Level == 0 {
+		d.level0Comp.Add(1)
+	} else {
+		d.nonLevel0Comp.Add(1)
+	}
+	d.activeComp++
+}
+
+func (d *DB) onCompactionEnd(info pebble.CompactionInfo) { //nolint:gocritic // Used by pebble's event listener
+	if d.activeComp == 1 {
+		d.compTime.Add(int64(time.Since(d.compStartTime)))
+	} else if d.activeComp == 0 {
+		panic("should not happen")
+	}
+	d.activeComp--
+}
+
+func (d *DB) onWriteStallBegin(b pebble.WriteStallBeginInfo) {
+	d.writeDelayStartTime = time.Now()
+}
+
+func (d *DB) onWriteStallEnd() {
+	d.writeDelayTime.Add(int64(time.Since(d.writeDelayStartTime)))
+}
+
+// collectMetrics periodically retrieves internal pebble counters and reports them to
+// the metrics subsystem.
+func (d *DB) StartMetricsCollection(ctx context.Context, refresh time.Duration) {
```

Comment:
Do you think it would be worth adding a test for this?

Reply:
I don't think we need to. It's not easy to add tests for them as we have to create an actual pebbleDB instead of relying on in-memory. If it goes wrong, it shouldn't affect the main application, and most likely we will detect the weird behaviors when we see it on Grafana.